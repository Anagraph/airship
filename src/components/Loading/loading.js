import React from 'react';
import styled, { keyframes } from 'styled-components';
import PropTypes from 'prop-types';

const rotate = keyframes`
  100% {
    transform: rotate(360deg);
  }
`;

const dash = keyframes`
  0% {
    stroke-dasharray: 1, 150;
    stroke-dashoffset: 0;
  }
  50% {
    stroke-dasharray: 90, 150;
    stroke-dashoffset: -35;
  }
  100% {
    stroke-dasharray: 90, 150;
    stroke-dashoffset: -124;
  }
`;

const StyledLoading = styled.div`
  height: ${(props) => `${props.size}px`};
  width: ${(props) => `${props.size}px`};

  svg {
    animation: ${rotate} 2s linear infinite;
  }

  circle {
    stroke: ${(props) => {
      console.log(props);
      return !!props.negative
        ? 'rgba(255, 255, 255, 1)'
        : 'rgba(44, 44, 44, 1)';
    }};
    stroke-linecap: round;
    animation: ${dash} 1.5s ease-in-out infinite;
    stroke-width: 4px;
  }
`;

const Loading = ({ size = 16, negative = false }) => {
  return (
    <StyledLoading size={size} negative={negative}>
      <svg viewBox="0 0 50 50">
        <circle cx="25" cy="25" r="20" fill="none" />
      </svg>
    </StyledLoading>
  );
};

Loading.propTypes = {
  size: PropTypes.oneOf([16, 32, 48, 64]),
  negative: PropTypes.bool
};

export default Loading;
