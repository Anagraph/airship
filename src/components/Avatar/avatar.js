import React, { Component } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Loading from '../Loading/loading';
import { colors } from '../../constants';

const StyledAvatar = styled.div`
  border-radius: 100%;
  width: ${(props) => `${props.size}px`};
  height: ${(props) => `${props.size}px`};
  background-repeat: no-repeat;
  background-size: cover;
  background-image: url(${(props) => props.url});
  background-color: ${colors.brand03};
`;
StyledAvatar.displayName = 'StyledAvatar';

const StyledLoading = styled.div`
  background-color: ${colors.ui02};
  border-radius: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  width: ${(props) => `${props.size}px`};
  height: ${(props) => `${props.size}px`};
`;
StyledLoading.displayName = 'StyledLoading';

class Avatar extends Component {
  state = {
    loaded: false
  };

  componentDidMount() {
    const { url } = this.props;
    const loaded = () => {
      this.setState({ loaded: true });
    };

    url === '' ? loaded() : fetch(url).then(loaded);
  }

  render() {
    const { size, url } = this.props;

    return this.state.loaded ? (
      <StyledAvatar size={size} url={url} />
    ) : (
      <StyledLoading size={size}>
        <Loading size={16} />
      </StyledLoading>
    );
  }
}

Avatar.propTypes = {
  size: PropTypes.oneOf([24, 32, 48]),
  url: PropTypes.string.isRequired
};

Avatar.defaultProps = {
  size: 24
};

export default Avatar;
