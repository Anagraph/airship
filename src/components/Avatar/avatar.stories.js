import React from 'react';
import { storiesOf } from '@storybook/react';
import { ThemeProvider } from 'styled-components';
import Avatar from './avatar';
import { theme } from '../../constants';

const CUSTOM_THEME = {
  ...theme,
  ui02: '#FABADA',
};

storiesOf('Avatar', module)
  .add('Default', () => (
    <Avatar url="https://avatars3.githubusercontent.com/u/1799254" />
  )).add('With custom theme', () => (
    <div>
      <ThemeProvider theme={CUSTOM_THEME}>
        <Avatar url="https://avatars3.githubusercontent.com/u/1799254" />
      </ThemeProvider>
    </div>
  ));
