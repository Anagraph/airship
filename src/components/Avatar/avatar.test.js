import React from 'react';
import Loading from '../Loading/loading';
import Avatar from './avatar';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

describe('render', () => {
  let originalFetch = global.fetch;

  beforeEach(() => {
    fetch.mockResponse(JSON.stringify({ image: 'wadus' }));
  });

  it('renders without crashing', () => {
    const component = shallow(<Avatar url="" />);
    expect(component).toMatchSnapshot();
  });

  it('renders with size', () => {
    const component = shallow(
      <Avatar
        size={32}
        url="https://avatars0.githubusercontent.com/u/1366843"
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('renders without image', () => {
    const component = shallow(<Avatar url="" />);
    expect(component).toMatchSnapshot();
  });

  it('renders loading state', () => {
    const component = shallow(<Avatar url="" />);
    component.setState({ loaded: false });
    expect(component.find('StyledLoading').length).toBe(1);
    expect(component.find('StyledAvatar').length).toBe(0);
    component.setState({ loaded: true });
    expect(component.find('StyledAvatar').length).toBe(1);
    expect(component.find('StyledLoading').length).toBe(0);
  });
});
