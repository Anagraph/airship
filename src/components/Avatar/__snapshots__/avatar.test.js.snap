// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders with size 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Avatar
    size={32}
    url="https://avatars3.githubusercontent.com/u/1799254"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Loading
        size={16}
      />,
      "size": 32,
      "theme": Object {
        "black": "#2C2C2C",
        "brand01": "#1785FB",
        "brand02": "#0F2D53",
        "brand03": "#47DB99",
        "complementaryColor": "#47DB99",
        "grey": "#E2E6E3",
        "primaryColor": "#1785FB",
        "secondaryColor": "#0F2D53",
        "support01": "#F3522B",
        "support02": "#FDB32B",
        "support03": "#80B622",
        "type01": "#2C2C2C",
        "type02": "#747474",
        "type03": "#BABABA",
        "type04": "#fff",
        "ui01": "#fff",
        "ui02": "#F5F5F5",
        "ui03": "#E2E6E3",
        "ui04": "#D1D5D7",
        "white": "#fff",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "size": 16,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Loading
          size={16}
        />,
        "size": 32,
        "theme": Object {
          "black": "#2C2C2C",
          "brand01": "#1785FB",
          "brand02": "#0F2D53",
          "brand03": "#47DB99",
          "complementaryColor": "#47DB99",
          "grey": "#E2E6E3",
          "primaryColor": "#1785FB",
          "secondaryColor": "#0F2D53",
          "support01": "#F3522B",
          "support02": "#FDB32B",
          "support03": "#80B622",
          "type01": "#2C2C2C",
          "type02": "#747474",
          "type03": "#BABABA",
          "type04": "#fff",
          "ui01": "#fff",
          "ui02": "#F5F5F5",
          "ui03": "#E2E6E3",
          "ui04": "#D1D5D7",
          "white": "#fff",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "size": 16,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`render renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Avatar
    size={24}
    url=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "size": 24,
      "theme": Object {
        "black": "#2C2C2C",
        "brand01": "#1785FB",
        "brand02": "#0F2D53",
        "brand03": "#47DB99",
        "complementaryColor": "#47DB99",
        "grey": "#E2E6E3",
        "primaryColor": "#1785FB",
        "secondaryColor": "#0F2D53",
        "support01": "#F3522B",
        "support02": "#FDB32B",
        "support03": "#80B622",
        "type01": "#2C2C2C",
        "type02": "#747474",
        "type03": "#BABABA",
        "type04": "#fff",
        "ui01": "#fff",
        "ui02": "#F5F5F5",
        "ui03": "#E2E6E3",
        "ui04": "#D1D5D7",
        "white": "#fff",
      },
      "url": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "size": 24,
        "theme": Object {
          "black": "#2C2C2C",
          "brand01": "#1785FB",
          "brand02": "#0F2D53",
          "brand03": "#47DB99",
          "complementaryColor": "#47DB99",
          "grey": "#E2E6E3",
          "primaryColor": "#1785FB",
          "secondaryColor": "#0F2D53",
          "support01": "#F3522B",
          "support02": "#FDB32B",
          "support03": "#80B622",
          "type01": "#2C2C2C",
          "type02": "#747474",
          "type03": "#BABABA",
          "type04": "#fff",
          "ui01": "#fff",
          "ui02": "#F5F5F5",
          "ui03": "#E2E6E3",
          "ui04": "#D1D5D7",
          "white": "#fff",
        },
        "url": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`render renders without image 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Avatar
    size={24}
    url=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "size": 24,
      "theme": Object {
        "black": "#2C2C2C",
        "brand01": "#1785FB",
        "brand02": "#0F2D53",
        "brand03": "#47DB99",
        "complementaryColor": "#47DB99",
        "grey": "#E2E6E3",
        "primaryColor": "#1785FB",
        "secondaryColor": "#0F2D53",
        "support01": "#F3522B",
        "support02": "#FDB32B",
        "support03": "#80B622",
        "type01": "#2C2C2C",
        "type02": "#747474",
        "type03": "#BABABA",
        "type04": "#fff",
        "ui01": "#fff",
        "ui02": "#F5F5F5",
        "ui03": "#E2E6E3",
        "ui04": "#D1D5D7",
        "white": "#fff",
      },
      "url": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "size": 24,
        "theme": Object {
          "black": "#2C2C2C",
          "brand01": "#1785FB",
          "brand02": "#0F2D53",
          "brand03": "#47DB99",
          "complementaryColor": "#47DB99",
          "grey": "#E2E6E3",
          "primaryColor": "#1785FB",
          "secondaryColor": "#0F2D53",
          "support01": "#F3522B",
          "support02": "#FDB32B",
          "support03": "#80B622",
          "type01": "#2C2C2C",
          "type02": "#747474",
          "type03": "#BABABA",
          "type04": "#fff",
          "ui01": "#fff",
          "ui02": "#F5F5F5",
          "ui03": "#E2E6E3",
          "ui04": "#D1D5D7",
          "white": "#fff",
        },
        "url": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
